### PR TITLE
ACP 77

### PR DESCRIPTION
**Pull request by sukantoraymond**
Base PR for all ACP 77 related PRs to be merged into

### CODE REVIEW

**Review by michaelkaplan13 — COMMENTED** [review]
👍 makes sense to me, though we may want to wait to merge this until the value is actually used for the validator manager contract deployment.

File: pkg/models/sidecar.go [hunk 1]
```diff
@@ -39,6 +39,7 @@ type Sidecar struct {
 	RunRelayer        bool
 	// SubnetEVM based VM's only
 	SubnetEVMMainnetChainID uint
+	ValidatorManagement     ValidatorManagementType
```
Comment (felipemadero) [inline pkg/models/sidecar.go:42]: we may want to add a boolean to decide if using SoV (Subnet Only Validators) or not.
and this validator management refers to SoV, probably SoValidatorManagement .
Maybe:
- UseSoV
- SoValidatorManagement

File: pkg/models/validator_management.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+type ValidatorManagementType string
+
+const (
+	ProofOfStake                 = "Proof Of Stake (Coming Soon)"
+	ProofOfAuthority             = "Proof Of Authority"
+	UndefinedValidatorManagement = "Undefined Validator Management"
+)
+
+func ValidatorManagementTypeFromString(s string) ValidatorManagementType {
```
Comment (felipemadero) [inline pkg/models/validator_management.go:13]: do we want to make this a member function of ValidatorManagementType?

Comment (felipemadero) [reply pkg/models/validator_management.go:13]: probably we also want a String(), so FromString(), ToString()

Comment (felipemadero) [reply pkg/models/validator_management.go:13]: also a IsUndefined() function?

File: tests/e2e/testcases/subnet/local/suite.go [hunk 3]
```diff
@@ -25,11 +25,6 @@ const (
 	subnetName       = "e2eSubnetTest"
 	secondSubnetName = "e2eSecondSubnetTest"
 	confPath         = "tests/e2e/assets/test_avalanche-cli.json"
-	stakeAmount      = "2000"
```
Comment (felipemadero) [inline tests/e2e/testcases/subnet/local/suite.go:28]: all of this was not used? :O

File: cmd/blockchaincmd/create.go [hunk 4]
```diff
@@ -101,6 +106,8 @@ configuration, pass the -f flag.`,
 	cmd.Flags().BoolVar(&createFlags.useWarp, "warp", true, "generate a vm with warp support (needed for teleporter)")
 	cmd.Flags().BoolVar(&createFlags.useTeleporter, "teleporter", false, "interoperate with other blockchains using teleporter")
 	cmd.Flags().BoolVar(&createFlags.useExternalGasToken, "external-gas-token", false, "use a gas token from another blockchain")
+	cmd.Flags().BoolVar(&createFlags.proofOfAuthority, "proof-of-authority", false, "use proof of authority for validator management")
```
Comment (felipemadero) [inline cmd/blockchaincmd/create.go:109]: do we need to add a comment that this is for new SoV ACP77? like (ACP77) or something, or (SOV)

File: cmd/blockchaincmd/create.go [hunk 5]
```diff
@@ -422,3 +436,43 @@ func checkInvalidSubnetNames(name string) error {
...
+// TODO: add explain the difference for different validator management type
+func promptValidatorManagementType(
+	app *application.Avalanche,
+	sidecar *models.Sidecar,
+) error {
+	proofOfAuthorityOption := "Proof of Authority"
+	proofOfStakeOption := "Proof of Stake"
+	explainOption := "Explain the difference"
+	if createFlags.proofOfStake {
+		sidecar.ValidatorManagement = models.ValidatorManagementTypeFromString(proofOfStakeOption)
```
Comment (felipemadero) [inline cmd/blockchaincmd/create.go:449]: use the constant at models here

File: cmd/blockchaincmd/create.go [hunk 6]
```diff
@@ -422,3 +436,43 @@ func checkInvalidSubnetNames(name string) error {
...
+		return nil
+	}
+	if createFlags.proofOfAuthority {
+		sidecar.ValidatorManagement = models.ValidatorManagementTypeFromString(proofOfAuthorityOption)
```
Comment (felipemadero) [inline cmd/blockchaincmd/create.go:453]: use the constant at models here

File: cmd/blockchaincmd/create.go [hunk 7]
```diff
@@ -422,3 +436,43 @@ func checkInvalidSubnetNames(name string) error {
 	}
 	return nil
 }
+
+// TODO: add explain the difference for different validator management type
+func promptValidatorManagementType(
+	app *application.Avalanche,
+	sidecar *models.Sidecar,
+) error {
+	proofOfAuthorityOption := "Proof of Authority"
```
Comment (felipemadero) [inline cmd/blockchaincmd/create.go:445]: you can use the constant string here from models

File: cmd/blockchaincmd/create.go [hunk 8]
```diff
@@ -422,3 +436,43 @@ func checkInvalidSubnetNames(name string) error {
...
+	proofOfStakeOption := "Proof of Stake"
```
Comment (felipemadero) [inline cmd/blockchaincmd/create.go:446]: you can use the constant string from models here

File: cmd/blockchaincmd/create.go [hunk 9]
```diff
@@ -422,3 +436,43 @@ func checkInvalidSubnetNames(name string) error {
...
+	explainOption := "Explain the difference"
+	if createFlags.proofOfStake {
+		sidecar.ValidatorManagement = models.ValidatorManagementTypeFromString(proofOfStakeOption)
+		return nil
+	}
+	if createFlags.proofOfAuthority {
+		sidecar.ValidatorManagement = models.ValidatorManagementTypeFromString(proofOfAuthorityOption)
+		return nil
+	}
+	options := []string{proofOfAuthorityOption, proofOfStakeOption, explainOption}
```
Comment (felipemadero) [inline cmd/blockchaincmd/create.go:456]: you need also to ask if the user wants to use or not SoV (that is the new ACP77 or the previous subnets).
or have SoV as a default but at least add a flag for it (most probably have a prompt?)

Comment (felipemadero) [reply cmd/blockchaincmd/create.go:456]: this before asking for validator management protocol for ACP77

File: pkg/models/validator_management.go (hunk 2 above)
Comment (felipemadero) [reply pkg/models/validator_management.go:13]: not agree, but lets move on

**Review by felipemadero — COMMENTED** [review]
LGTM as a base branch

**Review by arturrez — APPROVED** [review]

**Comment by meaghanfitzgerald** [comment]
🚀

**Pull request merged**